server/src/main/java/org/elasticsearch/search/diversification/ResultDiversificationRetrieverBuilder.java

```diff
@@ -0,0 +1,313 @@
+        Mapper mapper = ctx.getMappingLookup().getMapper(diversificationField);
+        if (mapper instanceof DenseVectorFieldMapper == false) {
+            throw new IllegalArgumentException("[" + diversificationField + "] is not a dense vector field");
+        }
+
+        if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR)) {
+            diversificationContext = new MMRResultDiversificationContext(
+                diversificationField,
+                lambda == null ? DEFAULT_LAMBDA_VALUE : lambda,
+                rankWindowSize,
+                (DenseVectorFieldMapper) mapper,
+                indexVersion,
+                queryVector == null ? null : new VectorData(queryVector),
+                null
+            );
+        } else {
+            // should not happen
+            throw new IllegalArgumentException("Unknown diversification type [" + diversificationType + "]");
+        }
+
+        return this;
+    }
+
+    @Override
+    protected SearchSourceBuilder finalizeSourceBuilder(SearchSourceBuilder sourceBuilder) {
+        return super.finalizeSourceBuilder(sourceBuilder).docValueField(diversificationField);
+    }
+
+    @Override
+    protected RankDoc[] combineInnerRetrieverResults(List<ScoreDoc[]> rankResults, boolean explain) {
+        // must have the combined result set
+        assert rankResults.size() == 1 : "ResultDiversificationRetrieverBuilder must have a single result set";
+        assert diversificationContext != null : "diversificationContext should be set before combining results";
+
+        ScoreDoc[] scoreDocs = rankResults.getFirst();
+        if (scoreDocs == null || scoreDocs.length == 0) {
+            // might happen in the case where we have no results
+            return new RankDoc[0];
+        }
+
+        assert scoreDocs[0] instanceof RankDocWithSearchHit : "expected results to be of type RankDocWithSearchHit";
+
+        // gather and set the query vectors
+        // and create our intermediate results set
+        RankDoc[] results = new RankDoc[scoreDocs.length];
+        Map<Integer, VectorData> fieldVectors = new HashMap<>();
+        for (int i = 0; i < scoreDocs.length; i++) {
+            RankDocWithSearchHit asRankDoc = (RankDocWithSearchHit) scoreDocs[i];
+            results[i] = asRankDoc;
+
+            var field = asRankDoc.hit().getFields().getOrDefault(diversificationField, null);
+            if (field != null) {
+                var fieldValue = field.getValue();
+                if (fieldValue instanceof float[]) {
+                    fieldVectors.put(asRankDoc.doc, new VectorData((float[]) field.getValue()));
+                } else if (fieldValue instanceof byte[]) {
+                    fieldVectors.put(asRankDoc.doc, new VectorData((byte[]) field.getValue()));
+                }
+            }
+        }
+        diversificationContext.setFieldVectors(fieldVectors);
+
+        try {
+            if (diversificationType.equals(DIVERSIFICATION_TYPE_MMR)) {
```

Comment:
You shouldn't need to do this check here, just call `diversify` from the parent class.